### PR TITLE
PERF: groupby.any/all with object dtype

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

```
from asv_bench.benchmarks.groupby import GroupByMethods

params = dict(dtype="object", method="all", application="direct", ncols=1) 

b = GroupByMethods()
b.setup(**params)

%timeit b.time_dtype_as_field(**params)

# 881 µs ± 20.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  -> main
# 216 µs ± 1.15 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  -> PR
```

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1774,22 +1774,13 @@ def _bool_agg(self, val_test: Literal["any", "all"], skipna: bool):
...
-                    func = np.vectorize(
-                        lambda x: bool(x) if not isna(x) else True, otypes=[bool]
-                    )
-                    vals = func(vals)
-                else:
-                    vals = vals.astype(bool, copy=False)
-
-                vals = cast(np.ndarray, vals)
+                vals = vals.copy()
+                vals[isna(vals)] = True
```
Comment (phofl) [inline pandas/core/groupby/groupby.py:1780]: Can you add a copy that we will re-compute a mask later on? Hard to follow why setting = True is valid if you don't know that.

Additionally, would it be faster to check if isna(vale).any() before making a copy?

Comment (lukemanley) [reply pandas/core/groupby/groupby.py]: Added a comment and skipped the copy when possible. Seems to save a bit more more (~5%).

File: doc/source/whatsnew/v2.0.0.rst [hunk 2]
```diff
@@ -752,6 +752,7 @@ Removal of prior version deprecations/changes
 Performance improvements
 ~~~~~~~~~~~~~~~~~~~~~~~~
 - Performance improvement in :meth:`.DataFrameGroupBy.median` and :meth:`.SeriesGroupBy.median` and :meth:`.GroupBy.cumprod` for nullable dtypes (:issue:`37493`)
+- Performance improvement in :meth:`.DataFrameGroupBy.all`, :meth:`.DataFrameGroupBy.any`, :meth:`.SeriesGroupBy.all`, and :meth:`.SeriesGroupBy.any` for object dtype (:issue:`#####`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.0.0.rst:755]: Can use PR # here assuming there is no issue.

**Review by rhshadrach — APPROVED** [review]
lgtm

File: doc/source/whatsnew/v2.0.0.rst (hunk 2 above)
Comment (lukemanley) [reply doc/source/whatsnew/v2.0.0.rst]: updated, thx

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
Thx @lukemanley
